persistent/cPersistence.c

```diff
@@ -609,17 +611,41 @@ pickle___setstate__(PyObject *self, PyObject *state)
 
         PyDict_Clear(*dict);
 
-        i = 0;
-        while (PyDict_Next(state, &i, &d_key, &d_value)) {
-            /* normally the keys for instance attributes are
-               interned.  we should try to do that here. */
-            if (NATIVE_CHECK_EXACT(d_key)) {
-                Py_INCREF(d_key);
-                INTERN_INPLACE(&d_key);
-                Py_DECREF(d_key);
+        if (PyDict_CheckExact(state))
+        {
+            i = 0;
+            while (PyDict_Next(state, &i, &d_key, &d_value)) {
+                /* normally the keys for instance attributes are
+                   interned.  we should try to do that here. */
+                if (NATIVE_CHECK_EXACT(d_key)) {
+                    Py_INCREF(d_key);
+                    INTERN_INPLACE(&d_key);
+                    Py_DECREF(d_key);
+                }
+                if (PyObject_SetItem(*dict, d_key, d_value) < 0)
+                    return NULL;
+            }
+        }
+        else
+        {
+            /* can happen that not a built-in dict is passed as state
+               fall back to iterating over items, instead of silently
+               failing with PyDict_Next */
+            items = PyMapping_Items(state);
```

Comment:
Please check for NULL. Which means an exception has been raised.
Take care to call Py_DECREF on `items` when they are not used. Otherwise memory leak.